### ISSUE TITLE
feature req: pass through Webpack "magic comments" inside import()

### ISSUE THREAD

**Issue by privatenumber** [evanw/esbuild#309]
## Feature request
I'm noticing comments are stripped from transpiled code even when minification is not enabled. Wondering if this can be something we opt into instead--can this be a minification option?

### Use-case
In [esbuild-loader](https://github.com/egoist/esbuild-loader), this is problematic because Webpack uses [magic comments](https://webpack.js.org/api/module-methods/#magic-comments) as hints for build configuration but esbuild strips them out:

```diff
-main: () => import(
-/* webpackChunkName: "chunk-name" */
-'../chunk-name'
-),
+main: () => import("../chunk-name")
```

I saw there was another use-case for it in https://github.com/evanw/esbuild/issues/294

### API proposal

Maybe something like this?
```
--minify-strip-comments="all"
--minify-strip-comments="non-license"
```

**Comment by evanw** [comment]
This has been brought up in the past as well (e.g. #221). Implementing comment preservation is not simple. There is no "comment stripping pass" in esbuild to disable. Comments are completely ignored at the moment, so doing this means a lot of extra work to special-case comment preservation into lots of individual places in the AST. It gets really tricky to do this when comments can be embedded in between any two tokens.

Right now esbuild doesn't preserve the token stream at all but to do this correctly I think you would have to change the AST into some form of token spanning tree, which is completely different than the current approach. It's also tricky to preserve comments through various syntax transformations. Given that esbuild is focused on bundling and speed, I consider general-purpose code formatting like this to be out of scope. There are other tools for this that already work quite well.

However, I can special-case certain comments into the AST. I think it could be reasonable to special-case these webpack "magic comments" in this specific place. That can just always be done without any additional configuration necessary. Keep in mind that esbuild won't be interpreting the contents of these comments, just passing them through unmodified.

**Comment by privatenumber** [comment]
Thanks for the insightful response!

I see the challenge now -- esbuild would have to create a concrete syntax tree and maintain careful positioning concerns while making optimizations, which is a high cost just for comments.

Seeing you also added special-cases for `PURE` and some comment patterns, I'm wondering if it would make sense to pass in regular expressions to preserve certain comments.

**Comment by evanw** [comment]
> Seeing you also added special-cases for `PURE` and some comment patterns, I'm wondering if it would make sense to pass in regular expressions to preserve certain comments.

Pure comments aren't preserved as comments. They mean something specific to the compiler that has to do with dead code elimination, and they can only appear on CallExpression and NewExpression AST nodes. They are represented as a boolean in the AST:

https://github.com/evanw/esbuild/blob/b3ff3306f01913210453c3fb751d09d6e7b401f9/internal/ast/ast.go#L453-L461

And then new comments are generated for these nodes when they are printed:

https://github.com/evanw/esbuild/blob/b3ff3306f01913210453c3fb751d09d6e7b401f9/internal/printer/printer.go#L1290-L1296

There are no comment strings to filter even if a regex was passed in. Obviously this could be implemented by switching the AST over to a general-purpose comment preservation tree. But I consider code formatting to be out of scope for esbuild, so I am not planning on making this change to the AST.

**Comment by privatenumber** [comment]
Gotcha. Thanks again for the careful explanations.

Special-casing the Webpack magic comments idea sounds like it should solve the problem I'm tackling.

**Comment by evanw** [comment]
This is in the newly-released version 0.6.15.

**Comment by privatenumber** [comment]
Amazing!! Thank you so much @evanw
